pkg/flag/pairlist.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flag
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/spf13/pflag"
+)
+
+// PairList is a flag value supporting a list of key=value pairs
+// It will optionally validate the supplied keys and / or values
+// It would parse something like "--arg foo=bar --arg baz=bar"
+type PairList struct {
+	Pairs       map[string]string
+	permissible map[string](map[string]struct{}) // map of allowed keys and, optionally, allowed values
+	typeName    string
+}
+
+var _ pflag.Value = (*PairList)(nil)
+
+// EMPTY is a small shortcut for the empty map
+var EMPTY = map[string]string{}
+
+// NewParList initializes a new pair list
+// If permissiblePairs is not empty, then it will validate keys and values. For
+// every key in permissiblePairs, if the value is an empty list, then any value
+// is allowed. If there is a list, then only those values are allowed for that.
+// for example, { "a": {"1", "two", "cheese"}, "b" : {}} would accept
+// a=1 and any b, but no other keys
+// defaults is a list of default parameters
+func NewPairList(permissiblePairs map[string][]string, defaults map[string]string) *PairList {
+	pl := &PairList{
+		Pairs:       make(map[string]string),
+		permissible: make(map[string]map[string]struct{}),
+		typeName:    "PairList",
+	}
+
+	// invert the list of allowed vals for fast lookup
+	for key, vals := range permissiblePairs {
+		pl.permissible[key] = make(map[string]struct{})
+
+		for _, val := range vals {
+			pl.permissible[key][val] = struct{}{}
+		}
+	}
+
+	for key, val := range defaults {
+		err := pl.SetOne(key, val)
+		if err != nil {
+			panic(err)
```

Comment:
What's going on with this one? Are we on track to get it into 1.15?

On 9 September 2016 at 14:00, Sergiusz Urbaniak <notifications@github.com>
wrote:

> In pkg/flag/pairlist.go
> <https://github.com/coreos/rkt/pull/3161#discussion_r78175257>:
>
> > +		typeName:    "PairList",
> > +	}
> > +
> > +	// invert the list of allowed vals for fast lookup
> > +	for key, vals := range permissiblePairs {
> > +		pl.permissible[key] = make(map[string]struct{})
> > +
> > +		for _, val := range vals {
> > +			pl.permissible[key][val] = struct{}{}
> > +		}
> > +	}
> > +
> > +	for key, val := range defaults {
> > +		err := pl.SetOne(key, val)
> > +		if err != nil {
> > +			panic(err)
>
> @squeed <https://github.com/squeed> well, it's not a pure initializer
> then ;-)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/rkt/pull/3161/files/48bab072184558e8570bcbae5e3e04a96ca47ec6#r78175257>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACewNxPNo1lp0_MsjR1a208WtmM1bE8Dks5qoVhmgaJpZM4JzuUZ>
> .
>